provider/pkg/googleclient/credentials.go

```diff
@@ -0,0 +1,171 @@
+	if c.AccessToken != "" {
+		contents, _, err := pathOrContents(c.AccessToken)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Error loading access token")
+		}
+		token := &oauth2.Token{AccessToken: contents}
+
+		if c.ImpersonateServiceAccount != "" {
+			opts := []option.ClientOption{
+				option.WithTokenSource(oauth2.StaticTokenSource(token)),
+				option.ImpersonateCredentials(c.ImpersonateServiceAccount, c.ImpersonateServiceAccountDelegates...),
+				option.WithScopes(c.Scopes...),
+			}
+			creds, err := transport.Creds(ctx, opts...)
+			if err != nil {
+				return nil, err
+			}
+			return creds, nil
+		}
+
+		glog.V(9).Info("Authenticating using configured Google JSON Access Token...")
+		glog.V(9).Infof("  -- Scopes: %s", c.Scopes)
+
+		return &google.Credentials{
+			TokenSource: staticTokenSource{oauth2.StaticTokenSource(token)},
+		}, nil
+	}
+
+	if c.Credentials != "" {
+		contents, _, err := pathOrContents(c.Credentials)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error loading credentials")
+		}
+		if c.ImpersonateServiceAccount != "" {
+			opts := []option.ClientOption{
+				option.WithCredentialsJSON([]byte(contents)),
+				option.ImpersonateCredentials(c.ImpersonateServiceAccount, c.ImpersonateServiceAccountDelegates...),
+				option.WithScopes(c.Scopes...),
+			}
+			creds, err := transport.Creds(ctx, opts...)
+			if err != nil {
+				return nil, err
+			}
+
+			glog.V(9).Info("Authenticating using configured Google JSON Credentials and Impersonate Service Account...")
+			glog.V(9).Infof("   -- Scopes: %s", c.Scopes)
+			return creds, nil
+		}
+		creds, err := google.CredentialsFromJSON(ctx, []byte(contents), c.Scopes...)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to parse credentials from %q", contents)
+		}
+
+		glog.V(9).Info("Authenticating using configured Google JSON Credentials...")
+		glog.V(9).Infof("   -- Scopes: %s", c.Scopes)
+		return creds, nil
+	}
+
+	if c.ImpersonateServiceAccount != "" {
+		opts := option.ImpersonateCredentials(c.ImpersonateServiceAccount, c.ImpersonateServiceAccountDelegates...)
+		creds, err := transport.Creds(ctx, opts, option.WithScopes(c.Scopes...))
+		if err != nil {
+			return nil, err
+		}
+
+		glog.V(9).Info("Authenticating using configured Impersonate Service Account...")
+		glog.V(9).Infof("   -- Scopes: %s", c.Scopes)
+		return creds, nil
+	}
+
+	glog.V(9).Info("Authenticating using DefaultClient...")
+	glog.V(9).Infof("   -- Scopes: %s", c.Scopes)
+
+	defaultTS, err := google.DefaultTokenSource(context.Background(), c.Scopes...)
+	if err != nil {
+		return nil, fmt.Errorf("Attempted to load application default credentials since neither `credentials` nor `access_token` was set in the provider block.  No credentials loaded. To use your gcloud credentials, run 'gcloud auth application-default login'.  Original error: %w", err)
+	}
+	return &google.Credentials{
+		TokenSource: defaultTS,
+	}, err
+}
+
+// TODO: Why do we need this?
```

Comment:
Isn't it used in `getCredentials` above? Or what is the question?

Reply:
> Isn't it used in `getCredentials` above? Or what is the question?

My specific question was around this comment: `staticTokenSource is used to be able to identify static token sources without reflection.`
I didn't see any introspection on the token source - wonder if this a hold over from the TF provider?

Reply:
Yes, probably

Reply:
I removed this type for now since we don't have great use for it.